components/brave_perf_predictor/browser/perf_predictor_tab_helper.cc

```diff
@@ -0,0 +1,144 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_perf_predictor/browser/perf_predictor_tab_helper.h"
+
+#include "brave/components/brave_perf_predictor/browser/third_parties.h"
+#include "brave/components/brave_perf_predictor/browser/third_party_extractor.h"
+#include "brave/components/brave_perf_predictor/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+namespace brave_perf_predictor {
+
+PerfPredictorTabHelper::PerfPredictorTabHelper(
+    content::WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  VLOG(3) << "Created PerfPredictorTabHelper";
+  if (web_contents->GetBrowserContext()->IsOffTheRecord()) {
+    return;
+  }
+
+  ThirdPartyExtractor* extractor = ThirdPartyExtractor::GetInstance();
+  if (!extractor->is_initialized()) {
+    extractor->load_entities(static_third_party_config);
+  }
+  bandwidth_predictor_ = std::make_unique<BandwidthSavingsPredictor>(extractor);
+#if BUILDFLAG(BRAVE_P3A_ENABLED)
+  bandwidth_tracker_ = std::make_unique<P3ABandwidthSavingsTracker>(
+      user_prefs::UserPrefs::Get(web_contents->GetBrowserContext()));
+#endif
+}
+
+PerfPredictorTabHelper::~PerfPredictorTabHelper() = default;
+
+void PerfPredictorTabHelper::DidStartNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle || !handle->IsInMainFrame() || handle->IsDownload()) {
+    return;
+  }
+  // Gather prediction of the _previous_ navigation
+  if (navigation_id_ != 0) {
+    RecordSavings();
+  }
+}
+
+void PerfPredictorTabHelper::ReadyToCommitNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle || !handle->IsInMainFrame() || handle->IsDownload()) {
+    return;
+  }
+  VLOG(3) << "Ready to commit navigation";
+  // Reset predictor state when we're committed to this navigation
+  if (bandwidth_predictor_) {
+    bandwidth_predictor_->Reset();
+  }
+}
+
+void PerfPredictorTabHelper::DidFinishNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle || !handle->IsInMainFrame() || !handle->HasCommitted() ||
+      handle->IsDownload()) {
+    return;
+  }
+
+  main_frame_url_ = handle->GetURL();
+}
+
+void PerfPredictorTabHelper::RecordSavings() {
+  if (bandwidth_predictor_ && web_contents()) {
+    uint64_t savings = (uint64_t)bandwidth_predictor_->predict();
+    VLOG(3) << "Saving computed bw saving = " << savings;
```

Comment:
this seems like a more useful log statement than the one above because it gives specific information about perf predictor state vs navigation state in the previous one